Document how to bin/scratch launched items into Grafana (#10382)
MaterializeInc/materialize: doc/developer/scratch.md

Pass `--dry-run` if you want to see what instances `bin/scratch destroy` would
destroy without actually destroyin them.

### Grafana integration

> The `materialized` process always exposes metrics at via its primary port's http server on the

Reviewer: ```suggestion
The `materialized` process always exposes metrics at its primary port's HTTP server on the
```

> prometheus-standard `/metrics` path, but the deployed server needs to be configured correctly for

Reviewer: ```suggestion
prometheus-standard `/metrics` path, but the scratch instance needs to be configured correctly for
```

(Wasn't clear to me whether "server" referred to the just-mentioned `materialized` process or the scratch instance; I think this is what you meant so I figured I'd make it explicit.)

> our Prometheus server to actually scrape the metrics and thereby expose them to Grafana. You

Reviewer 2: very gentle bump to remove the dangling "You"

The tl;dr is that you must configure mzcompose with `--preserve-ports` and the EC2 instance with

> the `"scrape_benchmark_numbers": "true"` tag. Thus a bare-minimum grafana-integrate config looks

Reviewer: ```suggestion
the `"scrape_benchmark_numbers": "true"` tag. Thus a bare-minimum Grafana-integrated config looks
```

like:

```javascript
{
    "launch_script": "bin/mzcompose --preserve-ports ..<remainder of args>"
    // .. snip config ..
    "tags": {
        "scrape_benchmark_numbers": "true"
    }
}
```

Read on for more details and some other items that can be configured.

#### Prometheus config

> There are three tags that configure our Prometheus' integration, and if you are running inside
> mzcompose it must also be configured correctly.
> 
> The three Prometheus tags are:

Reviewer: ```suggestion
There are three tags on the EC2 instance that configure our Prometheus integration:
```

(I think it makes more sense not talk about mzcompose here; you've already talked about it in general in the overview section above and you've got a dedicated section for it below!)

* `scrape_benchmark_numbers`: must be set to exactly the string `"true"` in order for Prometheus
  to observe the instance.
* `purpose`: is used as a filter in the Grafana UI. You can use this to group all your instances
  (e.g. set it to `myname-debugging`) or set it to `load-test` or `benchmark`.

> * `test` is both used by scratch to determine which test to execute, and displays as an additional

Reviewer 2: wait so just to double check - if the value of test doesn't match whats actually run in the launch script will that be a problem? I couldn't tell if the `test` field was purely informational and it seems like no?

Author: Test is used to determine what test to automatically run by the scratch script. If it is invalid then nothing will be run automatically.

Reviewer: > Test is used to determine what test to automatically run by the scratch script.

Hrm, are you sure? Where are you seeing that? Am I just missing it?

Author: Ah, no I must have been remembering the way it used to work, or just wrong.

  filter inside of the Grafana UI.

So the minimum scratch config to get metrics into Grafana looks like:

```javascript
{
    // .. snip general config ..
    "tags": {
        "scrape_benchmark_numbers": "true"
    }
}
```

And a slightly more complete one could be:

```javascript
{
    // .. snip general config ..
    "tags": {
        "scrape_benchmark_numbers": "true",
        "purpose": "bwm-debugging",
        "test": "chbench"
    }
}

{
    // .. snip general config ..
    "tags": {
        "scrape_benchmark_numbers": "true",
        "purpose": "bwm-debugging",
        "test": "billing"
    }
}
```

> #### MzCompose config

Reviewer: ```suggestion
#### mzcompose config
```

> Prometheus only looks for materialize metrics on port 6875. The canonical way to ensure that
> Materialize is available on port 6875 on a host is to pass the `--preserve-ports` argument to
> mzcompose.

Reviewer: ```suggestion
Prometheus only looks for Materialize metrics on port 6875. The canonical way to ensure that
Materialize is available on port 6875 on a host is to pass the `--preserve-ports` argument to
mzcompose. (Without this flag, mzcompose chooses a random host port for Materialize, which
will be unknown to Prometheus.)
```

(Just added a small note about what goes wrong if you don't specify `--preserve-ports`.)
